tests/testthat/test-renderer1-knit-print.R

```diff
@@ -1,168 +1,172 @@
+## https://github.com/rstudio/rmarkdown/issues/587#issuecomment-168437646
+## @yihui says "Do not use the output_dir argument of render()"
+rmarkdown::render(index.file)
+remDr$refresh()
+Sys.sleep(3)
+html <- getHTML()
+
+test_that("knit_print.animint renders five x axis titles", {
+  nodes <- getNodeSet(html, "//text[@class='xtitle']")
+  value.vec <- sapply(nodes, xmlValue)
+  expected.vec <-
+    c("first plot with color legend",
+      "second plot with color legend",
+      "non-interactive plot",
+      "position",
+      "segments")
+  expect_identical(value.vec, expected.vec)
+})
+
+test_that("segments and breakpoints are rendered", {
+  seg.list <- getNodeSet(html, '//g[@class="geom3_segment_signal"]//line')
+  expect_equal(length(seg.list), 6)
+  break.list <- getNodeSet(html, '//g[@class="geom4_vline_signal"]//line')
+  expect_equal(length(break.list), 5)
+})
+
+test_that("svg id property is unique", {
+  svg.list <- getNodeSet(html, "//svg")
+  attr.mat <- sapply(svg.list, xmlAttrs)
+  id.counts <- table(attr.mat["id",])
+  expect_true(all(id.counts==1))
+})
+
+all.list <- getNodeSet(html, "//*")
+id.na.vec <- sapply(all.list, function(e){
+  attr.vec.or.null <- xmlAttrs(e)
+  if("id" %in% names(attr.vec.or.null)){
+    attr.vec.or.null[["id"]]
+  }else{
+    NA
+  }
+})
+## In HTML, all values are case-insensitive
+## http://www.w3schools.com/tags/att_global_id.asp
+lower.id.vec <- tolower(id.na.vec)
+id.counts <- table(lower.id.vec)
+(not.unique <- id.counts[1 < id.counts])
+test_that("id property is unique over entire page", {
+  expect_equal(length(not.unique), 0)
+})
+
+test_that("id must contain at least one character", {
+  expect_true(all(0 < nchar(names(id.counts))))
+})
+
+test_that("id must not contain any space characters", {
+  expect_false(any(grepl(" ", names(id.counts))))
+})
+
+## function to extract all circles from an HTML page
+get_circles <- function(html=getHTML()) {
+  plot.names <- c("plot1top", "plot1bottom")
+  count.vec <- c()
+  for(i in seq_along(plot.names)){
+    xpath <- sprintf("//div[@id='%s']//circle[@class='geom']", plot.names[[i]])
+    circle.list <- getNodeSet(html, xpath)
+    count.vec[[i]] <- length(circle.list)
+  }
+  count.vec
+}
+
+get_elements <- function(id){
+  a<- runtime_evaluate(script=sprintf("div = document.getElementById('%s');div.getElementsByClassName('show_hide_selector_widgets')[0]", as.character(id)),return.value=TRUE)
+  b<- runtime_evaluate(script=sprintf("div = document.getElementById('%s');div.getElementsByClassName('show_hide_selector_widgets')[0]", as.character(id)),return.value=TRUE)
+  show_hide <- runtime_evaluate(script =sprintf("div = document.getElementById('%s');div.getElementsByClassName('table.legend tr.label_variable')[0]", as.character(id)), return.value=TRUE)
+  widget <-runtime_evaluate(script=sprintf("div = document.getElementById('%s');div.getElementsByClassName('table.legend tr.label_variable')[1]", as.character(id)),return.value=TRUE)
```

Comment:
please use a for loop or helper function instead of repeated JS code

Reply:
You mean something like this?
```R
runtime_evaluate_helper <- function(class_name, id, list_num) {
  script_template <- "div = document.getElementById('%s');div.getElementsByClassName('%s')['%s']"
  script <- sprintf(script_template, as.character(id), class_name, list_num)
  return runtime_evaluate(script = script, return.value = TRUE)
}
```

Reply:
yes a few functions like this would be great

Reply:
> yes a few functions like this would be great

I have made another helper function